BUG: Allow boolean subtract in histogram

AetherUnbound opened the pull request:
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->
This is one way to address #12107. I'm happy to implement this differently if there might be a better approach!

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

AetherUnbound commented:
That azure error doesn't seem to be at all related to this PR.

charris commented:
Close/open to trigger azure retest.

AetherUnbound commented:
This looks good to go!

charris commented on numpy/lib/tests/test_histograms.py:
```diff
@@ -249,6 +249,10 @@ def test_unsigned_monotonicity_check(self):
 
+        # gh-12107
+        # Shouldn't raise an error
+        hist, edges = np.histogram([True, False])
```
You should probably also test with `np.bool_`, which is the numpy boolean type, different from the python version. You can get the type with `np.bool_(1)`

AetherUnbound replied:
@charris good point, added.

eric-wieser commented:
I think a better fix would be to case the input data to `uint8` if it's `bool`, right at the beginning of the function - that way you'll fix the bin width estimators

I'm still not sure if this makes sense vs just requiring the caller to do the cast themselves.

AetherUnbound commented:
@eric-wieser Okay, I think that's a good call. I can also add a warning about the type conversion.

AetherUnbound commented:
This now does the conversion at the start, and I've fleshed out the tests a bit to make sure that it raises a warning as well as gives the output that was previously expected.

eric-wieser commented on numpy/lib/histograms.py:
```diff
@@ -220,6 +220,14 @@ def _hist_bin_auto(x):
     a = np.asarray(a)
+
+    # Ensure that the array is a "subtractable" dtype
+    if a.dtype in [bool, np.bool_]:
```
Just use `a.dtype == np.bool_`.

Using `==` on a dtype just casts the non-dtype into a dtype too.

eric-wieser commented on numpy/lib/tests/test_histograms.py:
```diff
@@ -141,6 +141,35 @@ def test_f32_rounding(self):
+
+        # Should also raise a warning on numpy booleans
+        with assert_raises(RuntimeWarning):
+            hist, edges = np.histogram([np.bool_(1), np.bool_(1), np.bool_(0)])
```
These two cases are indistinguishable, because histogram starts with `asarray`

eric-wieser commented:
Worth noting that even in 1.14, `np.histogram([True, True, False], bins='auto')` raises an error.

AetherUnbound commented:
Hmm, is this something that should be back ported? Or, since that functionality didn't work in 1.14, should it just be left as is?

charris commented on numpy/lib/tests/test_histograms.py:
```diff
@@ -141,6 +141,25 @@ def test_f32_rounding(self):
+        # the warnings to get the actual outputs
+        sup = suppress_warnings()
+        sup.filter(RuntimeWarning)
+        with sup:
```
It would be better to write this as
```
with suppress_warnings() as sup:
    sup.filter(RuntimeWarning)
```
to conform to the common usage.

charris replied:
You can also check the warnings here.

charris replied:
This would be better overall, if
```
# Reference integer histogram
￼a = np.array([1, 1, 0], dtype=np.uint8)
￼int_hist = np.histogram(a)

assert_warns(RuntimeWarning):
    hist = np.histogram([True, True, False])'
assert_array_equal(hist, int_hist)
```

AetherUnbound replied:
I'm happy to change this, I was just copying a pattern I found further down in the file: https://github.com/numpy/numpy/blob/7c5a3f4f9135af6ebe33dcd3242ef3e6db115cd6/numpy/lib/tests/test_histograms.py#L270

AetherUnbound replied:
Additionally (unless I'm doing something wrong), attempting to retrieve variables from the `assert_warns` call doesn't work:
```
    def test_bool_conversion(self):
        # gh-12107
        # Should raise an warning on booleans
        with assert_raises(RuntimeWarning):
            hist, edges = np.histogram([True, True, False])
>       print(hist)
E       UnboundLocalError: local variable 'hist' referenced before assignment
```

AetherUnbound committed:
BUG: Allow boolean subtract in histogram




Convert bool to uint at start, rather than attempt a XOR


Only check type against np.bool_


Refactor warnings check

AetherUnbound commented:
@charris I've modified the warning assertions to be less redundant and more specific.

mattip commented:
Thanks @AetherUnbound

AetherUnbound commented:
Thank you @charris and @eric-wieser for the assistance and guidance!
